main/autogen/autoloader.cc

```diff
@@ -400,17 +407,83 @@ void AutoloadWriter::write(const core::GlobalState &gs, const AutoloaderConfig &
                            UnorderedSet<std::string> &toDelete, const DefTree &node) {
     string name = node.root() ? "root" : node.name().show(gs);
     string filePath = join(path, fmt::format("{}.rb", name));
-    FileOps::writeIfDifferent(filePath, node.renderAutoloadSrc(gs, alCfg));
+    if (!alCfg.packagedAutoloader || !node.root()) {
+        FileOps::writeIfDifferent(filePath, node.renderAutoloadSrc(gs, alCfg));
+    }
     toDelete.erase(filePath);
     if (!node.children.empty()) {
         auto subdir = join(path, node.root() ? "" : name);
         if (!node.root() && !FileOps::dirExists(subdir)) {
             FileOps::createDir(subdir);
         }
         for (auto &[_, child] : node.children) {
-            write(gs, alCfg, subdir, toDelete, *child);
+            if (node.root() && alCfg.packagedAutoloader) {
+                // in a packaged context, we want to make sure that these constants are also put in their packages
+                ENFORCE(child->qname.package);
+                auto namespaceName = child->qname.package->show(gs);
+                // this package name will include the suffix _Package on the end, and we want to remove that
+                constexpr int suffixLen = 8; // strlen("_Package")
+                string packageName = namespaceName.substr(0, namespaceName.size() - suffixLen);
+                auto pkgSubdir = join(subdir, packageName);
```

Comment:
I think they're generated as though they're checked by Sorbet, but I for one am okay if we don't type-check them, so I think it's fine if they're grouped in an alternate package grouping.

It's worth noting: the alternative is generating them into a local `build` directory for each package, in which case they _would_ be grouped, but there'd also be `build` dirs scattered throughout the project. This would be a bit trickier to generate: we'd lose the property that all the autoloader files would be available in a nice flattened way (i.e. instead of `build/autoload/Package_Name/Source` it would now be in `lib/wherever/my/package/is/defined/build/Source`), but it _would_ let us continue type-checking the generated files.

@ngroman, do you have an opinion here?

Reply:
I like this approach (discussed on Slack) Sorbet already ignores all the autoload output. Keeping it all in one directory makes that easier.

Reply:
Discussed in Slack; so long as Sorbet continues to ignore these files then everything is fine.